src/renderer/atlas/BackendD3D.cpp

```diff
@@ -452,14 +452,22 @@ void BackendD3D::_recreateCustomShader(const RenderingPayload& p)
             if (error)
             {
                 LOG_HR_MSG(hr, "%.*hs", static_cast<int>(error->GetBufferSize()), static_cast<char*>(error->GetBufferPointer()));
+                if (p.warningCallback)
+                {
+                    //to handle compile time errors
+                    const char* errorMessage = static_cast<char*>(error->GetBufferPointer());
+                    std::wstring errorMessageWString(errorMessage, errorMessage + strlen(errorMessage));

```

Comment:
This isn't compatible with Unicode text. It effectively compiles down to a
```cpp
for (int i = 0; i < len; i++) {
    dst[i] = src[i];
}
```
In other words, it blindly casts the `char`s to `wchar_t`s.

Instead, add the following include to the top of the file:
```cpp
#include "../../types/inc/ColorFix.hpp" <-- already exists
#include "../../types/inc/convert.hpp"  <-- new
```

and use something like this:
```cpp
const std::string_view msgA{ static_cast<const char*>(error->GetBufferPointer()), error->GetBufferSize() };
const auto msgW = ConvertToW(CP_ACP, msgA);
```

After thinking about this for a bit, I think it's fine to remove the two `LOG_HR` calls. They don't really do anything anyway. So you'd basically end up with something like this?
```cpp
if (SUCCEEDED(hr))
{
    THROW_IF_FAILED(p.device->CreatePixelShader(blob->GetBufferPointer(), blob->GetBufferSize(), nullptr, _customPixelShader.addressof()));
    // ...
}
else if (p.warningCallback)
{
    if (error)
    {
        const std::string_view msgA{ static_cast<const char*>(error->GetBufferPointer()), error->GetBufferSize() };
        const auto msgW = ConvertToW(CP_ACP, msgA);
        p.warningCallback(hr, msgW);
    }
    else
    {
        p.warningCallback(hr, p.s->misc->customPixelShaderPath);
    }
}
```

...or maybe something similar. Up to you! :)

Reply:
Hmm I think we also need to replace the `hr` in the 2nd `warningCallback` call with `D2DERR_SHADER_COMPILE_FAILED` because otherwise the `parameter` (= `customPixelShaderPath`) won't show up in the popup.

Reply:
Hello! I implemented the changes you mentioned, unicode compatibility should not be a problem now.
Thank you for your time; I hope this makes sense and won't sound dumb.

About the `hr` in the 2nd `warningCallback`....
Possible values of `hr` when compilation fails but it is not due to the compiler generating an error (ie. the `error` is `NULL`):
`HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)` when the filename is incorrect
`HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND)` when the path is incorrect
`E_ACCESSDENIED` when the file cannot be accessed

This is why the 2nd `warningCallback` has `hr` instead of `D2DERR_SHADER_COMPILE_FAILED`, so that appropriate warnings are generated for the aforementioned errors.
TermControl.cpp _RendererWarning function's switch case:
![image](https://github.com/microsoft/terminal/assets/104504393/7622a033-92b6-4ccb-9cb4-25be2610b35b)

If we pass `D2DERR_SHADER_COMPILE_FAILED`, it won't generate the desired warnings.

Warnings generated after the most recent change:
![image](https://github.com/microsoft/terminal/assets/104504393/d433f74f-8119-4226-b531-253ccc7670f8)
![image](https://github.com/microsoft/terminal/assets/104504393/389b43b0-7e0f-4581-a922-d8c38704a084)
![image](https://github.com/microsoft/terminal/assets/104504393/3acfec16-0242-44ae-9a66-1a6960aaa9bf)

Reply:
That makes sense!